### DISCUSSION TITLE
posts/new-scenarios/

### DISCUSSION THREAD

**Discussion by giscus** [polyclimate/polyclimate.org#7] [Announcements]
# posts/new-scenarios/

Creation and Assessment of Climate Scenarios.

https://polyclimate.org/posts/new-scenarios/

**Comment by znicholls** [comment]
Nice. On the scenario process, you know my very confused thoughts :)

On the living database. GH actions is a very interesting thought that could work super well. You could also do an RCMIP style approach (where we had people upload from their computer, and some crunching was done locally before uploading). That has the benefit of having people get results locally first, with the downside they have to install first. Of course, the best solution is probably to do both.

The one that really cooked us in AR6 was when we had to re-run the entire database because we updated the emulators. For that, GH actions doesn't really work and you quickly get too big for git (because now your data size doesn't grow one scenario at a time, it instantaneously increases by the size of all scenarios). I think we'd need a better data storage solution, but there are lots out there that don't come with too many headaches (I don't think).

**Reply by rgieseke** [reply]
Ye, I think there are many possibilities! Also the storage doesn't _have_ to be on GitHub. Could also upload to any server or storage engine. I'm not sure about the actual processing requirements of *climate-assessment*. When running locally I had to reduce the number of parallel runs on my laptop as otherwise it would crash.
One could also have a mix of locally run and using 'some' institute's cluster. Every scenario is submitted RCMIP style with assessment and when the emulator version changes a giant PR needs to be prepared on a cluster.

**Reply by znicholls** [reply]
Ye nice I like it

**Comment by rgieseke** [comment]
> You could also do an RCMIP style approach (where we had people upload from their computer, and some crunching was done locally before uploading). That has the benefit of having people get results locally first, with the downside they have to install first. Of course, the best solution is probably to do both.

I really like the RCMIP example. It brought together quite a community on different platforms, so it's a nice example that such an approach could work for others as well.

There has been criticism that there are barriers to entry for modellers not at the core of the IAM groups, a GitHub-based approach would probably create barriers, too, but on the other hand it's always possible to help each other and built different tools for different people, e.g.
- command line tool for upload from terminal
- PR based workflow, could also work directly from web interface
- a web interface where an Excel file is dropped

**Reply by jkikstra** [reply]
Hey, nice! 

Just to pop in briefly: many of these things have been implemented in the IIASA Scenario Explorer architecture. So far in discussions we've had with some in the IAM community, bits that are still to solve are funding (for the coding support & maintenance, vetting), governance & ensuring inclusive and broad engagement structures including scientific coordination (selecting/finding the people that will steer different groups that implement vetting, other add-on modules, maintenance). 

The scenario explorer backend used to actually have a MAGICC6 setup running automatically, but now has been updated to (be able to) run climate-assessment automatically; on an IIASA computer. The outputs are then made available in the Scenario Explorer when the run is done (~20mins later i guess on that computer), which you can visualise directly in the web view, or download the data.
Scenario Explorer instances (generally 1 internal per project, and then 1 public once the data development is ready) have GitHub instances visible to the collaborators. Daniel Huppmann leads this team.
Modelling teams (at the moment just those with access) can upload their data as an Excel file. 
PRs are used for registering models (think especially: how should IAM regions be aggregated to standardized regions). 

Downloading the data can also be done directly via the pyam package, as long as you have the login data and access.
Uploading via pyam is perhaps something to be considered too (if Daniel Huppmann and other colleagues don't already use/have something better).

A more PR-centric workflow I like the idea of; mainly for the reason that it could be a great way to include a vetting procedure before new uploads end up in a database.

**Reply by znicholls** [reply]
I think these are good thoughts, but I do hesitate a bit on some things:

- I think it would be great to build whatever we're looking at here so it is decoupled from IIASA architecture. Locking ourselves into architecture decisions that were made 15 years ago doesn't seem the most obvious route to me (better to make something decoupled, then push that into IIASA, then something coupled to IIASA that can never be used elsewhere)
- For some data processing, pyam is great. For anything climate model related that actually uses all the ensemble members, it doesn't work (and rightly so, because it's built for a different domain). Again, having some more abstract downloading would be a better long-term solution I think as we need solutions that work across multiple domains (and something like pooch is so abstract and well done that we'd be better off just using that, then wrapping e.g. pyam over the top rather than forcing everything through pyam)
- I think we also all know how flaky the climate assessment package is. Lots of work to be done there so we can use it in future reliably without massive development headaches

**Reply by rgieseke** [reply]
Thanks for chiming in, @jkikstra, that's quite useful context! 

On the *climate-assessment* package, it might be flaky, @znicholls, I too had to tweak the number of parallel runs to prevent crashes, but it is a massive achievement. Great documentation and example notebooks allowed me to get toy scenarios running quickly!

**Reply by znicholls** [reply]
Oh ye running stuff it's great for! I'm just already squirming at the idea of updating things for whatever the next set of runs is (heaps of refactoring to do so that we can actually swap out components sensibly!).

**Reply by rgieseke** [reply]
Got it, just wanted to add some positive-ness to the flaky-ness! When looking how and where things were defined I too wondered how hard it would be to swap things out or change defaults.

**Reply by znicholls** [reply]
I think very, we made a highly coupled system (good for users, bad for us) rather than a decoupled system with a coupled layer on top (giving users an easy experience but also making further configuration in future easier for us). The refactor will be ok, just takes some time.

**Reply by GlenPetersCICERO** [reply]
On the "barrier to entry" issue. The discussion here implies the climate assessment is essential, & it is the main point of the entire exercise. If I take the ENGAGE scenarios, they were constructed based on cumulative CO2, & I could analyse them that way (without needing a climate assessment). Other scenarios may may look at NDCs, or CBAMs, or policy mixes, or... and the primary point of comparison or assessment may not be the climate assessment. I think the "barrier to entry" issue also relates to the 1) feeling by some that there is more interesting analysis to do then putting scenarios in temperature boxes and 2) the feeling their model is somehow inferior because it doesn't try to model everything. Just imagine a quiet PhD student developing a really cool model, but then looks at the database with 100s of scenarios each with 100s of variables and 10s of regions from huge modelling groups. 

These issues come back to the purpose. Is the purpose of the scenario database to do climate assessments? Is it to more easily assess the literature? Etc. It may make sense to split the climate assessment into an alternative database (that is easily linked). You could think of the IIASA Scenario Explorer as a separate tool for displaying the database & interacting with the database, but not where people go to get the database. Vetting does not (and should not) be done behind closed doors, but different groups may develop cool tools to do diverse types of vetting. Some activities make sense to centralise (database curation, cleaning, maintenance), some activities make sense to diversify. What is the database and what is an independent tool that uses the database? What tools should we encourage people to independently develop, with the best tool(s) winning? Etc

**Reply by rgieseke** [reply]
Very interesting aspects, thanks for sharing @GlenPetersCICERO!
I think what you describe aligns well with a potentially more decentralised on GitHub. One improvement could also be that research projects/IAM papers share their results immediately in a somewhat standardized repository format.

These could then be integrated in larger repos doing vetting, climate assessment or other analyses.

**Reply by GlenPetersCICERO** [reply]
Yes, I think a standard format can be used for sharing (e.g., IAMC format). An institution with support can quality control and maintain a core database (e.g., IIASA). The research community would develop tools and do analysis. I think some diversity and competition is important to push the research frontier. There many be a competing groups doing climate assessments, or maybe they self organise (e.g., RCMIP), but there could be one or more secondary databases with climate assessments. Over time the research community will hopefully determine the pros and cons of different approaches, and then come AR7, etc, there will be a more robust community driven literature that IPCC authors then assess.

**Comment by rgieseke** [comment]
I have implemented an experimental test repo with the `climate-assessment` pipeline running on GitHub. In a Pull Request a scenario is added (already infilled with all gases), in the GitHub Action the assessment runs, some example plots are added to the Pull Request and the output data is committed back. After review it can then be added. 

Repo: https://github.com/polyclimate/polyclimate-scenario-db
Example Pull Request: https://github.com/polyclimate/polyclimate-scenario-db/pull/2
